### ISSUE TITLE
Training and validation of an image classifier pipeline

### ISSUE THREAD

**Issue by elboyran** [DynaSlum/SateliteImaging#37]
Follow example [Image Category Classifier](https://nl.mathworks.com/help/vision/examples/image-category-classification-using-bag-of-features.html?requestedDomain=www.mathworks.com) but also use own scripts.
Use some default choices like SURF like 80% of the strongest features. multi-class SVM, but also own parameters:

- vocabulary sizes of 10, 20 and 50 

- SURF locations 'Detector' and 'Grid' (default settings GridStep is [8 8] and the BlockWidth is [32 64 96 128 (> 100??)] )

Apply on data-set 6, 100px = 80m.

- [x] make script skeleton

- [x] Balance categories

- [x] Looping over some parameters

- [ ] Implement cross-validation

- [x] Make better output

- [x] Publish

- [x] Determine the best classifier from the tested vocabulary sizes and SURF location points

- [x] Run with best classifier options and save the model, publish.

- [x] Share with partners

**Comment by elboyran** [comment]
Creating image data store for tile size: 100 pixels = 80 meters.

imds = 

  ImageDatastore with properties:

       Files: {
              ' ...\Mumbai_P4_R1C1_3_clipped_rgb_tile_sr1001er1100sc1241ec1340.tif';
              ' ...\Mumbai_P4_R1C1_3_clipped_rgb_tile_sr1001er1100sc1291ec1390.tif';
              ' ...\Mumbai_P4_R1C1_3_clipped_rgb_tile_sr1001er1100sc1381ec1480.tif'
               ... and 31887 more
              }
      Labels: [BuiltUp; BuiltUp; BuiltUp ... and 31887 more categorical]
    ReadSize: 1
     ReadFcn: @readDatastoreImage

      Label       Count
    __________    _____

    BuiltUp       10800
    NonBuiltUp    11090
    Slum          10000

-----------------------------------------------------------------
Splititng the datastore into Train and Test datastores with fractionTrain: 70%
Training datastore: 
      Label       Count
    __________    _____

    BuiltUp       7560 
    NonBuiltUp    7763 
    Slum          7000 

Testing datastore: 
      Label       Count
    __________    _____

    BuiltUp       3240 
    NonBuiltUp    3327 
    Slum          3000 

-----------------------------------------------------------------
Creating features (BoVW) using vocabulary size 10 for the Train datastore keeping the 70% strongest features

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: NonBuiltUp
* Image category 3: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 22323 images...done. Extracted 661134 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 80466.
** Using the strongest 80466 features from each of the other image categories.

* Using K-Means clustering to create a 10 word visual vocabulary.
* Number of features          : 241398
* Number of clusters (K)      : 10

* Initializing cluster centers...100.00%.
* Clustering...completed 40/100 iterations (~0.73 seconds/iteration)...converged in 40 iterations.

* Finished creating Bag-Of-Features

-----------------------------------------------------------------
Training an image categiry classifier

Training an image category classifier for 3 categories.
--------------------------------------------------------
* Category 1: BuiltUp
* Category 2: NonBuiltUp
* Category 3: Slum

* Encoding features for 22323 images...done.

* Finished training the category classifier. Use evaluate to test the classifier on a test set.

-----------------------------------------------------------------
Evaluating perfomance on the Training set

Evaluating image category classifier for 3 categories.
-------------------------------------------------------

* Category 1: BuiltUp
* Category 2: NonBuiltUp
* Category 3: Slum

* Evaluating 22323 images...done.

* Finished evaluating all the test sets.

* The confusion matrix for this test set is:


                          PREDICTED
KNOWN         | BuiltUp   NonBuiltUp   Slum   
----------------------------------------------
BuiltUp       | 0.75      0.09         0.16   
NonBuiltUp    | 0.14      0.67         0.19   
Slum          | 0.15      0.09         0.76   

* Average Accuracy is 0.73.

Evaluating perfomance on the Testing set

Evaluating image category classifier for 3 categories.
-------------------------------------------------------

* Category 1: BuiltUp
* Category 2: NonBuiltUp
* Category 3: Slum

* Evaluating 9567 images...done.

* Finished evaluating all the test sets.

* The confusion matrix for this test set is:


                          PREDICTED
KNOWN         | BuiltUp   NonBuiltUp   Slum   
----------------------------------------------
BuiltUp       | 0.76      0.09         0.16   
NonBuiltUp    | 0.13      0.68         0.19   
Slum          | 0.15      0.09         0.76   

* Average Accuracy is 0.73.

**Comment by elboyran** [comment]
To balance the categories use:

`minSetCount = min(tbl{:,2});` % determine the smallest amount of images in a category

% Use splitEachLabel method to trim the set.
`imds = splitEachLabel(imds, minSetCount, 'randomize');`

% Notice that each set now has exactly the same number of images.
`countEachLabel(imds)`

**Comment by elboyran** [comment]
Creating image data store for tile size: 100 pixels = 80 meters.

imds = 

  ImageDatastore with properties:

       Files: {
              ' ...\Mumbai_P4_R1C1_3_clipped_rgb_tile_sr1001er1100sc1241ec1340.tif';
              ' ...\Mumbai_P4_R1C1_3_clipped_rgb_tile_sr1001er1100sc1291ec1390.tif';
              ' ...\Mumbai_P4_R1C1_3_clipped_rgb_tile_sr1001er1100sc1381ec1480.tif'
               ... and 31887 more
              }
      Labels: [BuiltUp; BuiltUp; BuiltUp ... and 31887 more categorical]
    ReadSize: 1
     ReadFcn: @readDatastoreImage

      Label       Count
    __________    _____

    BuiltUp       10800
    NonBuiltUp    11090
    Slum          10000

-----------------------------------------------------------------
Balancing the datasets.

ans = 

      Label       Count
    __________    _____

    BuiltUp       10000
    NonBuiltUp    10000
    Slum          10000

Splititng the datastore into Train and Validation datastores with fractionTrain: 30%
Training datastore: 
      Label       Count
    __________    _____

    BuiltUp       3000 
    NonBuiltUp    3000 
    Slum          3000 

Testing datastore: 
      Label       Count
    __________    _____

    BuiltUp       7000 
    NonBuiltUp    7000 
    Slum          7000 

-----------------------------------------------------------------
Creating features (BoVW) using vocabulary size 10 for the Train datastore keeping the 70% strongest features

Creating Bag-Of-Features.
-------------------------
* Image category 1: BuiltUp
* Image category 2: NonBuiltUp
* Image category 3: Slum
* Selecting feature point locations using the Detector method.
* Extracting SURF features from the selected feature point locations.
** detectSURFFeatures is used to detect key points for feature extraction.

* Extracting features from 9000 images...done. Extracted 268891 features.

* Keeping 70 percent of the strongest features from each category.

* Balancing the number of features across all image categories to improve clustering.
** Image category 2 has the least number of strongest features: 31058.
** Using the strongest 31058 features from each of the other image categories.

* Using K-Means clustering to create a 10 word visual vocabulary.
* Number of features          : 93174
* Number of clusters (K)      : 10

* Initializing cluster centers...100.00%.
* Clustering...completed 37/100 iterations (~0.22 seconds/iteration)...converged in 37 iterations.

* Finished creating Bag-Of-Features

-----------------------------------------------------------------
Training an image categiry classifier

Training an image category classifier for 3 categories.
--------------------------------------------------------
* Category 1: BuiltUp
* Category 2: NonBuiltUp
* Category 3: Slum

* Encoding features for 9000 images...done.

* Finished training the category classifier. Use evaluate to test the classifier on a test set.

-----------------------------------------------------------------
Evaluating perfomance on the Training set

Evaluating image category classifier for 3 categories.
-------------------------------------------------------

* Category 1: BuiltUp
* Category 2: NonBuiltUp
* Category 3: Slum

* Evaluating 9000 images...done.

* Finished evaluating all the test sets.

* The confusion matrix for this test set is:


                          PREDICTED
KNOWN         | BuiltUp   NonBuiltUp   Slum   
----------------------------------------------
BuiltUp       | 0.81      0.06         0.14   
NonBuiltUp    | 0.15      0.63         0.22   
Slum          | 0.15      0.07         0.77   

* Average Accuracy is 0.74.

Evaluating perfomance on the Validation set

Evaluating image category classifier for 3 categories.
-------------------------------------------------------

* Category 1: BuiltUp
* Category 2: NonBuiltUp
* Category 3: Slum

* Evaluating 21000 images...done.

* Finished evaluating all the test sets.

* The confusion matrix for this test set is:


                          PREDICTED
KNOWN         | BuiltUp   NonBuiltUp   Slum   
----------------------------------------------
BuiltUp       | 0.81      0.05         0.14   
NonBuiltUp    | 0.15      0.63         0.22   
Slum          | 0.16      0.07         0.77   

* Average Accuracy is 0.74.

**Comment by elboyran** [comment]
## Conclusions:
1. BoVW training with different vocabulary sizes doesn't affect the number of features (?), but only the number of iterations. The number of features when using Detector locations are 104016 and using Grid is 4867200.

2. Training the classier (muli-class linear SVM) with the SURF features from the previous step 

3. Evaluation of the classifier's performance. The used measures are:

- Confusion matrix (for now normalized)

- accuracy

- sensitivity

- specificity

- precision

- recall

- Fscore


## **Main conclusions:**

- The performance on the training and testing sets is comparable.

- The 'Detector' interest point locations are better than using 'Grid' locations.

- Larger vocabulary size gives better results

**==>> From the tested the best is BoVW = 50 and 'Detector'**



For full deatils see C:\Projects\DynaSlum\Results\Classification3Classes\PerformanceComparision\html_classifier
